### ISSUE TITLE
unreachable warning when an unreachable!() macro is called

### ISSUE THREAD

**Issue by gilescope** [rust-lang/rust#79157]
I think we should suppress warnings that unreachable!() macros are unreachable. I had some nested logic with early returns and while at the moment it could not get to the end of the function, I could predict that some changes to the logic might miss a branch and then it would get to the end of the function - I wanted to catch that future mistake with an unreaachable!() macro. If we don't warn about unreachable unreachable!'s then I think that will make rust a safer language to code in as we can code more defensively.

```
warning: unreachable statement
    --> src/<snip>.rs:1298:5
     |
1190 |                         return None;
     |                         ----------- any code following this expression is unreachable
...
1298 |     unreachable!("Case missed. Should have early returned above.");
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unreachable statement
```

**Comment by jyn514** [comment]
@gilescope I'm curious that you describe this as coding defensively - what does the `unreachable!` give you that the lint doesn't?

**Comment by gilescope** [comment]
Here's an example of the kind of coding error I was trying to pick up:

```rust
if rats_nest {
    // coder refactors this section and forgets the else branch:
    if cond {
         return Z;
    }
} else {
    if let Some(_) = other {
        return Y;
    } else {
        return X;
    }
}
unreachable!("Shouldn't be able to get here due to early returns");
```

In the above if I miss coding the else then I'd like to catch that. We could achieve the same result by converting all the early returns to expressions and the compiler would then catch the missing case - I think that's nicer where possible, but if the return type was `()` then I don't think the compiler would help and an unreachable!() statement might be a reasonable approach?
